[IMP] website: keep showing dropzone message even if content inside

rdeodoo opened the pull request:
There is a very nice and clear system that helps the user to figure what are the available area to drop content inside, and what will they do: are they shared between products, or product specific etc.

The issue is that once you have dropped a snippet inside, that helper message is not shown anymore, but it's still not obvious which area is used for what, actually you have no clue which one is which, especially if you come back on the page later: at best remember there were 2 separate zones but you don't especially remember which one is which.

Keeping the message helps in that regard without any negative impact.

Note that making the message appear when there are already a snippet will work out of the box in the sense that the message will be duplicated and shown twice: one at the very bottom of the area and one at the very top.

task-3160416

For tracking purpose, here is how it looks like in actual 16.0:
| color | before | after |
| -- | -- | -- |
| white | ![image](https://user-images.githubusercontent.com/30048408/224693692-ea6111d1-429b-4fac-a9c3-b4295656d066.png) |![image](https://user-images.githubusercontent.com/30048408/224694899-f70ed5b1-675f-4e15-ade9-2d8513b58492.png) |
|odoo purple | ![image](https://user-images.githubusercontent.com/30048408/224693807-5540993e-a14f-42f9-8c74-fae4c34cff0b.png) |![image](https://user-images.githubusercontent.com/30048408/224694810-3e8b56dd-ac06-42e4-8848-fcada8de1896.png) |
| odoo green | ![image](https://user-images.githubusercontent.com/30048408/224693905-d5c8d0ff-6ae7-48a1-b56d-e572a02ffba4.png) | ![image](https://user-images.githubusercontent.com/30048408/224694863-2c3e19f1-5c77-4d58-86dc-da57069f824c.png)|
| gray | ![image](https://user-images.githubusercontent.com/30048408/224694004-f480899e-9ae3-441d-b006-1745ea1354d8.png) | ![image](https://user-images.githubusercontent.com/30048408/224694664-a0462485-e7e5-4457-afaa-0236075298ef.png) |
| black | ![image](https://user-images.githubusercontent.com/30048408/224694324-a1f0db75-693b-4124-bab1-c2c2b77466f7.png) | ![image](https://user-images.githubusercontent.com/30048408/224694737-046e1a4c-1d87-4d59-8039-401057cfe30f.png)|

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/111515).

rdeodoo commented:
@qsm-odoo On the fly PR I made a few weeks ago while PDE was asking me about the spec of the task: to reply I had to check and came up with the "imp".
I'll create a follow-up task afterward to harmonize texts between our apps, will be a functional job to go through each of them and write down spec.

qsm-odoo commented on addons/website/static/src/scss/website.edit_mode.scss:
```diff
@@ -81,14 +81,34 @@ $-editor-messages-margin-x: 2%;
             @extend %o-editor-messages;
-            padding: 112px 0px;
+
+            &:only-child {
+                padding: 112px 0px;
+            }
```
You now apply `&:only-child` to `#wrap:empty` which was not the case before, intended?

qsm-odoo replied:
> if its empty it can't have a only child

That not what `:only-child` means, it checks if you are the only-child yourself. So this would break a layout where `#wrap has a neighbor :wink:

qsm-odoo commented on addons/website/static/src/scss/website.edit_mode.scss:
```diff
@@ -81,14 +81,34 @@ $-editor-messages-margin-x: 2%;
+                padding: 112px 0px;
+            }
+
+            &:not(:only-child) {
+                padding: 0px;
+                &::before {
+                    font-size: 18px;
+                    line-height: 25px!important;
+                }
+            }
```
- This code has been copy-pasted (with a nice line break that there should be here too :wink:), can't we share it? Especially since I don't see why there is an exception for `#wrap` anymore?
- Is the padding 0 useful? (just curious)

rdeodoo replied:
Yes, it's actually to reset the padding which we override in `%o-editor-messages` which was initially 0 before that mixin kicks off: https://github.com/odoo-dev/odoo/blame/6840c928962deb8f9fb870528f2f8749c598a8d0/addons/web_editor/static/src/scss/wysiwyg_snippets.scss#L1954

But I have refactored that part (and will continue later) anyway, instead of erasing for all cases and restoring it in some cases only, I'll just replace it for the cases we want

qsm-odoo commented on addons/website/static/src/scss/website.edit_mode.scss:
```diff
@@ -81,14 +81,34 @@ $-editor-messages-margin-x: 2%;
+            }
         }
-        > .oe_drop_zone.oe_insert:not(.oe_vertical):only-child {
+        > .oe_drop_zone.oe_insert:not(.oe_vertical) {
```
It does not appear code-wise but a side effect of this PR:

Before:
![image](https://user-images.githubusercontent.com/10338094/223513845-7c573623-9b55-4ac2-b34c-5c8526b1a028.png)

After:
![image](https://user-images.githubusercontent.com/10338094/223513739-5ec3d4bb-5721-4ac6-a5b6-8df9cd6cae11.png)

=> Spacing around the dropzone. I visually like it (kinda the same to me) but functionally I find this very bad: the page "jumps" when toggling the dropzones because they now take spacing in the page (in the example above, the first snippet is now lower than before toggling the dropzones). I'd really prefer to keep the fact that toggling the dropzones just make them appear as an overlay of the page, without making snippets move around... and once we do that, I am not sure that displaying the text will look nice :grimacing: (to check)

cc @pierre-pde

pierre-pde replied:
@qsm-odoo  & @rdeodoo  IMO, the "jump" is not so problematic because, if you decide to drop the block there, there will be a jump.
But I understand that it's not ideal, let see how it looks like without the dropzone taking space.

qsm-odoo replied:
IMO, there is no relation at all between the "jump" post-drop and the "jump" pre-drop :grimacing: Here the problem is that you scroll somewhere to add a block, start dragging and the whole screen changes because blocks are being pushed towards the bottom by `~100px * number of snippets above them`.

IMO it's not "not ideal", it's a no-go. Here an example with some snippets (I used multiples number snippets) + a media list + a numbers + a banner:
![Screenshot from 2023-03-08 14-36-13](https://user-images.githubusercontent.com/10338094/223727456-47c08cea-0db3-449b-80e0-705ac68c3cce.png)

=> I want to add something between the number and the banner (location which is visible on the screen), I start dragging, I jump to this:
![Screenshot from 2023-03-08 14-36-26](https://user-images.githubusercontent.com/10338094/223727783-26185711-6832-47ca-92d1-d4c44e62a8be.png)

=> My location is gone I have to scroll (up? down?) to find it again.

Also small side-effect: the auto-scroll after drop is actually slightly flickering because of this too: the dropzones taking space are being removed during the smooth scroll making the animation not as smooth in some cases.

pierre-pde replied:
Ok, of course the effect that user is being pushed towards the bottom `by ~100px * number of snippets above them` is not acceptable, I agree

rdeodoo replied:
Note that it's not `by ~100px * number of snippets above them` but `by ~100px * number of visible at screen snippets above them`.
But indeed good catch it's really terrible in those proportions

qsm-odoo replied:
> Note that it's not by ~100px * number of snippets above them but by ~100px * number of visible at screen snippets above them.

Definitely not, cfr my screenshots above explaining the issue: there is jump of more than one screen height, while only 2-3 snippets are visible on the screen :wink:

rdeodoo replied:
Would you both mind having a look at the runbot and tell me what you think?
- Don't pay attention to the code
- Don't pay attention to the dropzone text style (color/bg color), it's a wip, I still have to find something which works on every color. It basically was an issue we already have as reported by the india testing: go to /forum and drag & drop snippet in the Welcome message, same issue.
- So just functionally about the 5 use cases:
  - Empty website.page in edit mode (no drag & drop)
  - Empty website.page with drag & drop
  - Non empty website.page with drag & drop
  - Empty product description with drag & drop
  - Non empty product description with drag & drop

If that sounds reasonable, I'll fine tune my ~crap~ wip

pierre-pde replied:
@rdeodoo , I had a look, the UI looks great to me.
I don't like the fact the placeholder has a big font-weight, but I know we're not supposed to judge style.

qsm-odoo replied:
@rdeodoo I closed my eyes and tested and it looks great :smile: ... joke aside I am honestly confused on what you want me to check: it's like before but with the text (which is a nice idea) but I can't judge the text style... so in the end... I can't say much right? :smile: 

Maybe just a suggestion: maybe there is no need to display the text for the standard use case? When the text is "Drag and drop a building block here" ? That way that would lighten the UI by a lot while still providing what you are trying to fix: being able to see which dropzones are common to some pages in some cases? And that would reduce the need to make this text perfectly visible in all cases?

rdeodoo replied:
I am not sure how we could differentiate a base case from another case tho?
We can of course differentiate the website.page case from the product case but even in the product case there is a kind of "base case" which is the product description field and a more complexe case with the "Share it here etc..".

qsm-odoo replied:
The default message is automatically added in JS as the `data-editor-message` attribute (something like that). That JS could also put something like `data-is-default-editor-message="true"` so you can differentiate in CSS.

rdeodoo commented:
@qsm-odoo @pierre-pde This one was harder than expected :grimacing: 

I didn't yet consider @qsm-odoo [suggestion](https://github.com/odoo/odoo/pull/111515#discussion_r1132257775) (waiting for reply) but apart from that I updated a cleaned and working version of the IMP.
You can have a look and tell me what you think, including style wise ;-)
It was a bit more nice with white background before, but it's way better now with pretty much any other bg color. I mentioned that imp too in the commit message and I updated the PR description with a comparison before/after so it's easier to track :hot_face:

qsm-odoo commented:
> It was a bit more nice with white background before, but it's way better now with pretty much any other bg color.

@rdeodoo Problem is that light background feels more important to me than any other. Although, it is indeed worse in your screenshot, but in practice, it seems better. I think there is a problem somewhere in the branch: I have dropzones with less opacity than others which creates the problem I think:
![image](https://user-images.githubusercontent.com/10338094/224735795-6674f507-f249-46cb-9939-02996ac27e4a.png)
=> Clearly in this screenshot (both over white background), the second dropzone is clearly readable... but the first one is not.

IMO, if all dropzones looked like the clearer one, it would be nice :+1: I would however still consider not showing the message for the default case.

Ultimately, I'll let @pierre-pde be the judge here :wink:

rdeodoo commented:
> I think there is a problem somewhere in the branch: I have dropzones with less opacity than others which creates the problem I think:

yeah.. It's a weird rule I didn't dare to remove because it seemed to be explicitly here for that:
```scss
.oe_structure_not_nearest .oe_drop_zone {
    &:before {
        opacity: 0.5;
        line-height: 35px !important;
    }
}
```
See https://github.com/odoo/odoo/commit/22a76c230bb22dfa42333b544f69cc72d8921a31

> Problem is that light background feels more important to me than any other.

Note that I am totally up for any other solution to make the text readable on every layout, but I will need someone else idea on this. I kinda tried a ton a solution and couldn't find something better than this (if we want to keep the opacity on those, which I think we should).
I spent already too much time on it and I am not able to find something better myself

qsm-odoo commented:
> (if we want to keep the opacity on those, which I think we should).

Why should we keep a visual difference between those dropzones? And why should it be about opacity if we do need to keep a visual difference?

qsm-odoo commented:
@rdeodoo Well yes, because those are the exact same points... you have `opacity: 1` on the ones that have no text readability issues and a different opacity for `oe_structure_not_nearest` as you said (`opacity: 0.5`) which IMO causes the text readability issue on my side.

So: why do you think we should keep the opacity difference since removing it seems to:
- Have a more consistent UI
- Less CSS rules
- Solve the white background issue we have in this PR, without breaking anything else (at least on my side?)

Took the time to test it out, seems nice (both same opacity, both on white background):
![image](https://user-images.githubusercontent.com/10338094/224757604-d9885cb5-7c36-4c62-bbdf-e0687d6d41ef.png)


I am probably missing something since you seem annoyed but ... tell me what :shrug:

rdeodoo commented:
> Well yes, because those are the exact same points...

No they are not, seems like you misunderstood me 2 unrelated points, but probably because I wrote `if we want to keep the opacity on those, which I think we should` and not `if we want to keep the background opacity on those, which I think we should` (**background** opacity).
My bad, was an important precision ^^

So let's start again:
The first thing I was referring to: there is a 50% opacity on the `oe_structure_not_nearest`. I don't like it and I don't think it makes sense, but it was coded on purpose in an explicit commit so I didn't remove it.
The second thing is that there is a background opacity on the dropzone messages, which is the reason why it is so hard to have a text which is readable with all colors. That's this one I said "we should keep it imo", it looks cool, otherwise there will be too much contrast between the background color of this message and the one from the element on which it applies. So I was saying that since we probably want to keep this one, I can't find a better way without impacting the white bg :)

> I am probably missing something since you seem annoyed but

Really not :p You reply was confusing the 2 unrelated points (because of my poor wording / miss of precision) so there wasn't much else to say apart from what I said :D

qsm-odoo commented:
Ok but from my side, I don't see how the background opacity is a problem given the font styles you used (like I said in my first message: it seems good expect for the dropzones which have an opacity rule (not background opacity)). Given the fact you used white letters, mixing that background with white will lead to the worse result and in that case, with:
- Keeping the background opacity rules as they are
- Removing the whole rule added by the specific commit you mentioned (with the opacity stuff)

-> we reach the screenshot I added in my previous message: https://user-images.githubusercontent.com/10338094/224757604-d9885cb5-7c36-4c62-bbdf-e0687d6d41ef.png and in that case, I'd say everything seems good? Let's do that? :)

rdeodoo commented:
Sure, that's what I wanted to do initially but seem like it was a solution to an UX pain point at the time.
Let's remove it and we will see later, everything should be easier to figure now with the many imp we made with the drag & drop anyway.

rdeodoo commented:
Done alongside this change: https://github.com/odoo/odoo/compare/bbb06f52e8ccc23ecfa154cbcbbe915ff654ee86..9ebd1fca6719b5a49e08506329d21bf4d9f3031a (completely removed the line-height tweaks).
Default dropzones will be a bit thinner (5px) which doesn't seems worth the line of code.
Even if someone was to purposely break the line-height, like 0.1, it would still work (made sure of that) even if it would be as tall as the text: I can restore a default one if judged necessary.

qsm-odoo commented:
I'll let @pierre-pde judge before checking this any further :+1:

pierre-pde commented:
@rdeodoo  & @qsm-odoo ,
I tested the readability on light, grey & dark bg, the text is always readable.
So I would say design achieve its goal :) 
There is one new behavior that is really not a big deal but a bit strange, the dropzone size changes when they are no building block on static page:
![image](https://user-images.githubusercontent.com/75257428/224954466-825b5d34-d1fd-415e-8349-16a29b7ba83f.png)
It's a detail though, I'm fine if you decide to leave as-is.

rdeodoo commented:
- Fixed @pierre-pde bug report (it was really not ok imo) by dropping the wizz helper effect: https://github.com/odoo/odoo/compare/9ebd1fca6719b5a49e08506329d21bf4d9f3031a..87a023181eb855db3a63c1fba57cec82420607e8 It's fine to do it, we improved the UX already to help figuring the drag & drop
- Added the request of @qsm-odoo about not showing the default text when there is multiple elements inside already with https://github.com/odoo/odoo/compare/87a023181eb855db3a63c1fba57cec82420607e8..b80db4cc556ccfa9ab43575e66d8c064af3ecf8d
You can see that part easily on an empty product description page, add one element in both shared and not shared area then start drag & drop, only the shared area will show the message

rdeodoo commented:
@qsm-odoo Looks like this one went to oblivion :D
Whenever you have time, it can be reviewed and/or merged.
The recent change you requested (second point in my previous message) was flagged as blocking by RD India, I let them know that it was expected and better that way :+1:

qsm-odoo commented on addons/website/static/src/components/wysiwyg_adapter/wysiwyg_adapter.js:
```diff
@@ -350,6 +350,7 @@ export class WysiwygAdapterComponent extends ComponentAdapter {
     _addEditorMessages() {
         const $wrap = this.$editable.find('.oe_structure.oe_empty, [data-oe-type="html"]');
         this.$editorMessageElement = $wrap.not('[data-editor-message]')
+                .attr('data-editor-message-default', true)
```
Meh, don't like giving a non-string to `attr` but I guess this will go away in a no-jquery refactoring at some point anyway :+1:

qsm-odoo reviewed (approved):
> Looks like this one went to oblivion :D

No, as you said, it was still in testing phase :wink:  But let's bypass it :+1:

Found 2 bugs while testing but apparently already there before:
- On an empty website.page -> see the pretty drag and drop message, drag a block inside, then remove it, the drag and drop message is not pretty anymore. Worsened by this PR because of removed padding but I guess this can wait a further bug fix in our drag and drop review later.
- /shop -> filter on a category -> edit -> drag and a snippet in top section and in category section, start dragging again -> the products are pushed to the bottom. Same: we'll review that in drag and drop review later. Those are not critical bugs anyway.

@robodoo r+
